### PR TITLE
Add support for gopls 'implements' functionality

### PR DESCRIPTION
**Pull request by woodstok**
Add support to use the textDocument/implementation gopls feature.
Add g:go_implements_mode so that users can choose between gopls and
guru.

### CODE REVIEW

File: autoload/go/lsp/message.vim [hunk 1]
```diff
@@ -189,7 +202,7 @@ endfunction
 
 function! go#lsp#message#ChangeWorkspaceFolders(add, remove) abort
   let l:addDirs = map(copy(a:add), function('s:workspaceFolder', []))
-  let l:removeDirs = map(copy(a:add), function('s:workspaceFolder', []))
+  let l:removeDirs = map(copy(a:remove), function('s:workspaceFolder', []))
```
Comment (woodstok) [inline autoload/go/lsp/message.vim]: This is a minor fix that is unrelated to the Implements functionality.

**Comment by woodstok** [comment]
In addition to the default Implements feature, 
https://go-review.googlesource.com/c/tools/+/219678/ is in the works which would bring a reverse Implements functionality. Hopefully that merges soon to gopls

File: autoload/go/list.vim [hunk 2]
```diff
@@ -151,6 +151,7 @@ let s:default_list_type_commands = {
       \ "GoTest":               "quickfix",
       \ "GoVet":                "quickfix",
       \ "GoReferrers":          "quickfix",
+      \ "GoImplements":         "quickfix",
```
Comment (bhcleek) [inline autoload/go/list.vim]: The default for `GoImplements` should be `locationlist`, because the information that it the list will contain is relative to a single buffer.

File: autoload/go/lsp.vim [hunk 3]
```diff
@@ -830,6 +830,31 @@ function! s:referencesHandler(next, msg) abort dict
   call call(a:next, [0, l:result, ''])
 endfunction
 
+" go#lsp#Implementations calls gopls to get the implementations to the identifier at line
+" and col in fname. handler should be a dictionary function that takes a list
+" of strings in the form 'file:line:col: message'. handler will be attached to
+" a dictionary that manages state (statuslines, sets the winid, etc.). handler
+" should take three arguments: an exit_code, a JSON object encoded to a string
+" that mimics guru's ouput for `what`, and third mode parameter that only
+" exists for compatibility with the guru implementation of SameIDs.
```
Comment (bhcleek) [inline autoload/go/lsp.vim]: (nit): `SameIDS` -> `implements`

There's no need for you to change this; I'm just leaving a note for myself to fix this after merging.

File: autoload/go/lsp.vim [hunk 4]
```diff
@@ -830,6 +830,31 @@ function! s:referencesHandler(next, msg) abort dict
...
+function! go#lsp#Implements(fname, line, col, handler) abort
+  call go#lsp#DidChange(a:fname)
+
+  let l:lsp = s:lspfactory.get()
+  let l:msg = go#lsp#message#Implementation(a:fname, a:line, a:col)
+
+  let l:state = s:newHandlerState('implements')
+
+  let l:state.handleResult = funcref('s:referencesHandler', [function(a:handler, [], l:state)], l:state)
```
Comment (bhcleek) [inline autoload/go/lsp.vim]: Just another note for myself to do later: `s:referencesHandler` should be refactored to call another function. This should call a new function `s:implementsHandler`, which calls that common function.

File: autoload/go/lsp.vim [hunk 5]
```diff
@@ -830,6 +830,31 @@ function! s:referencesHandler(next, msg) abort dict
...
+  let l:state.error = funcref('s:implementsErrorHandler', [function(a:handler, [], l:state)], l:state)
+  return l:lsp.sendMessage(l:msg, l:state)
+endfunction
+
+function! s:implementsErrorHandler(next, error) abort dict
+  call call(a:next, [-1, [a:error], ''])
```
Comment (bhcleek) [inline autoload/go/lsp.vim]: `-1` is ok, but `1` may be more consistent with other exit codes.

File: autoload/go/implements.vim [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+" or locationlist.
+function! s:parse_output(exit_val, output, title) abort
+  if a:exit_val
+    call go#util#EchoError(a:output)
+    return
+  endif
+
+  let errformat = ",%f:%l:%c:\ %m"
+  let l:listtype = go#list#Type("GoImplements")
+  call go#list#ParseFormat(l:listtype, errformat, a:output, '')
```
Comment (bhcleek) [inline autoload/go/implements.vim]: The last argument should be `a:title`.

**Comment by bhcleek** [comment]
@woodstok do you mind if I push to your branch before merging?

**Comment by woodstok** [comment]
@bhcleek Feel free

**Comment by woodstok** [comment]
> In addition to the default Implements feature,
> https://go-review.googlesource.com/c/tools/+/219678/ is in the works which would bring a reverse Implements functionality. Hopefully that merges soon to gopls

This has merged to gopls v0.3.3 

@bhcleek, just to clarify, the review comments were for your reference right? Or were you waiting for me to make the changes?

**Comment by bhcleek** [comment]
Thanks for checking in. I'm planning to make the changes, I've just been very busy lately

**Comment by bhcleek** [comment]
@woodstok given how this works now, I think it might also be a good replacement for `:GoCallees`. WDYT?

**Comment by woodstok** [comment]
> @woodstok given how this works now, I think it might also be a good replacement for `:GoCallees`. WDYT?

I am not sure if they map to the same functionality. And I can't seem to get my go guru to work properly in my projects, so can't verify either.

**Comment by bhcleek** [comment]
Guru works in GOPATH mode. For `GoCallees`, you may need to call `:GoGuruScope` with the scope that you want to evaluate.

**Comment by bhcleek** [comment]
Thank you for your patience @woodstok . Now that I have a few bugs off my plate, I'm going to focus on getting this merged.

I'll be pushing to your branch soon as we previously discussed.

**Comment by bhcleek** [comment]
Thank you again for the contribution 🙇

**Comment by woodstok** [comment]
> Thank you for your patience @woodstok . Now that I have a few bugs off my plate, I'm going to focus on getting this merged.
> 
> I'll be pushing to your branch soon as we previously discussed.

Thank you for the merge. I was away from my laptop for quite a while.
